src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -853,132 +874,185 @@ public void ProcessQueue(SocketAsyncContext context)
-                            if (_state == QueueState.Stopped)
+                            // We're the head of the queue
+                            if (op == _tail)
                             {
-                                Debug.Assert(_tail == null);
-                                Trace(context, $"Exit (stopped)");
+                                // No more operations 
+                                _tail = null;
                             }
                             else
                             {
-                                Debug.Assert(_state == QueueState.Processing, $"_state={_state} while processing queue!");
-                                Debug.Assert(_tail.Next == op, "Queue modified while processing queue");
+                                // Pop current operation and advance to next
+                                _tail.Next = op.Next;
+                            }
 
-                                if (op == _tail)
+                            // We're the first op in the queue.
+                            if (_state == QueueState.Processing)
+                            {
+                                // The queue has already handed off execution responsibility to us.
+                                // We need to dispatch to the next op.
+                                if (_tail == null)
                                 {
-                                    // No more operations to process
-                                    _tail = null;
                                     _state = QueueState.Ready;
                                     _sequenceNumber++;
-                                    Trace(context, $"Exit (finished queue)");
                                 }
                                 else
                                 {
-                                    // Pop current operation and advance to next
-                                    nextOp = _tail.Next = op.Next;
+                                    nextOp = _tail.Next;
+                                }
+                            }
+                            else if (_state == QueueState.Waiting)
+                            {
+                                if (_tail == null)
+                                {
+                                    _state = QueueState.Ready;
                                 }
                             }
                         }
-                    }
-                    else
-                    {
-                        // Check for retry and reset queue state.
-
-                        using (Lock())
+                        else
                         {
-                            if (_state == QueueState.Stopped)
+                            // We're not the head of the queue.
+                            // Just find this op and remove it.
+                            AsyncOperation current = _tail.Next;
+                            while (current.Next != op)
```

Comment:
Is there any risk here that the op may not be in the list?  If that could happen, this would be an infinite loop/hang.

Reply:
Only if our state is badly corrupt.  I could check for this, but if it happens then we'd probably want to FailFast.  I suppose that's slightly better than hanging, but not sure it's worth it.